### ISSUE TITLE
CFE is erasing typedefs with nullable types incorrectly

### ISSUE THREAD

**Issue by leafpetersen** [dart-lang/sdk#46518]
The specification of canonicalization for type literals with null safety is [here](https://github.com/dart-lang/language/blob/master/accepted/2.12/nnbd/feature-specification.md#type-literals) .  This specification was designed to ensure that type literals which were "spelled the same" compared equal and identical, regardless of whether they occurred in opted in or opted out code (this was considered important in order to support AngularDart dependency injection patterns).

It looks like the CFE (I think, but maybe the backends?) is being more aggressive and erasing all nullable types when canonicalizing typedefs.  Example:

opted_in.dart
```dart
typedef NullableIntF = int? Function();

void _check(Type t1, Type t2) {
  print("Opted in: identical($t1, $t2) == ${identical(t1, t2)}");
  print("Opted in: ($t1 == $t2) == ${t1 == t2}");
}

void checkOptedIn(Type t) {
  _check(t, NullableIntF);
}
```

test.dart
```dart
// @dart=2.8

import "opted_in.dart";

void main() {
  checkOptedIn(NullableIntF);
}
```

This code prints out:

```
Opted in: identical(() => int, () => int?) == false
Opted in: (() => int == () => int?) == false
```

In other words, the *same typedef* referred to by the *same name* is comparing neither identical nor equal if one reference is in an opted in library and another is in an opted out library.

I believe this is a bug per the spec, unless I'm missing something (and if I am, I think we need to resolve it in the spec and then fix it).

Note that this is likely a hard blocker for opening up AngularDart code migration internally.

cc @johnniwinther @stefantsov @grouma @leonsenft @davidmorgan

**Comment by grouma** [comment]
> Note that this is likely a hard blocker for opening up AngularDart code migration internally.

Note we were able to work around this problem by injecting an `OpaueToken` in lieu of the typedef by annotating the constructor with the token. It's not clear to me if that approach will work in all situations however and of course increases the burden of migration.

**Comment by stefantsov** [comment]
It can be that the CFE performs legacy erasure too early. I'm experimenting with one possible solution here: https://dart-review.googlesource.com/c/sdk/+/205523.

**Comment by leafpetersen** [comment]
Seems like there there's still a case that's being handled incorrectly here.  See the example below: my read of the spec [here](https://github.com/dart-lang/language/blob/master/accepted/2.12/nnbd/feature-specification.md#const-evaluation-and-canonicalization) is that these should be treated as identical, but evaluating this on the VM prints false.

cc @johnniwinther @grouma 

// opted in lib.dart
```
typedef NullableIntF = Future<int?> Function();

class OT<T extends Object> {
  const OT();

  @override
  String toString() {
    return "$runtimeType";
  }
}

const token = OT<NullableIntF>();
```

// opted out
```dart
// @dart=2.9
import "lib.dart";

const localToken = OT<NullableIntF>();
void main() {
  print(
      "identical($token, $localToken) evaluates to ${identical(token, localToken)}");
}
```
